builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+                                                mappedArgs[0],
+                                                passthroughArgs[0]
+                                           });
+        };
+
+        return CreateMapToI32(function, pValue, pIndex);
+    }
+}
+
+// =====================================================================================================================
+// Create a subgroup shufflexor call.
+Value* BuilderImplSubgroup::CreateSubgroupShuffleXor(
+    Value* const pValue,   // [in] The value to shuffle.
+    Value* const pMask,    // [in] The index to shuffle from.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    Value* const pMbcnt = CreateMbcnt(getInt64(UINT64_MAX));
+    Value* const pXor = CreateXor(pMbcnt, pMask);
+    return CreateSubgroupShuffle(pValue, pXor, instName);
+}
+
+// =====================================================================================================================
+// Create a subgroup shuffleup call.
+Value* BuilderImplSubgroup::CreateSubgroupShuffleUp(
+    Value* const pValue,   // [in] The value to shuffle.
+    Value* const pIndex,   // [in] The index to shuffle from.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    Value* const pMbcnt = CreateMbcnt(getInt64(UINT64_MAX));
+    Value* const pSub = CreateSub(pMbcnt, pIndex);
+    return CreateSubgroupShuffle(pValue, pSub, instName);
+}
+
+// =====================================================================================================================
+// Create a subgroup shuffledown call.
+Value* BuilderImplSubgroup::CreateSubgroupShuffleDown(
+    Value* const pValue,   // [in] The value to shuffle.
+    Value* const pIndex,   // [in] The index to shuffle from.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    Value* const pMbcnt = CreateMbcnt(getInt64(UINT64_MAX));
+    Value* const pAdd = CreateAdd(pMbcnt, pIndex);
+    return CreateSubgroupShuffle(pValue, pAdd, instName);
+}
+
+// =====================================================================================================================
+// Create a subgroup clustered reduction.
+Value* BuilderImplSubgroup::CreateSubgroupClusteredReduction(
+    GroupOperation groupOperation, // The group operation.
+    Value* const   pValue,         // [in] An LLVM value.
+    Value* const   pClusterSize,   // [in] The cluster size.
+    const Twine&   instName)       // [in] Name to give final instruction.
+{
+    if (getContext().GetGfxIpVersion().major >= 8)
+    {
+        // Start the WWM section by setting the inactive lanes.
+        Value* pResult = CreateSetInactive(pValue, CreateGroupOperationIdentity(groupOperation, pValue->getType()));
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(2)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPMov(pResult, DPPCtrl::DPP_QUAD_PERM1032, 0xf, 0xf, 0)), pResult);
```

Comment:
please add comments for 0xf, 0xa, 0x8 in CreateDPPMov parameters

Reply:
Done!